ksp_plugin_adapter/reference_frame_selector.cs

```diff
@@ -0,0 +1,134 @@
+﻿#define HAS_SURFACE
+#define HAS_BODY_CENTRED_ALIGNED_WITH_PARENT
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace principia {
+namespace ksp_plugin_adapter {
+
+static class CelestialExtensions {
+  public static bool is_leaf(this CelestialBody celestial) {
+    return celestial.orbitingBodies.Count == 0;
+  }
+
+  public static bool is_root(this CelestialBody celestial) {
+    return celestial.orbit == null;
+  }
+}
+
+class ReferenceFrameSelector {
+
+  private enum FrameType {
+    BODY_CENTRED_NON_ROTATING,
+#if HAS_SURFACE
+    SURFACE,
+#endif
+    BARYCENTRIC_ROTATING,
+#if HAS_BODY_CENTRED_ALIGNED_WITH_PARENT
+    BODY_CENTRED_ALIGNED_WITH_PARENT
+#endif
+  }
+
+  Dictionary<CelestialBody, bool> expanded_;
```

Comment:
A `Dictionary` with value `bool` is generally an awkward way to represent a set.  Consider using HashSet<> or SortedSet<>.